lib/stdlib/src/timer.erl

```diff
@@ -26,129 +26,195 @@
-      Module :: module(),
-      Function :: atom(),
-      Arguments :: [term()],
-      TRef :: tref(),
-      Reason :: term().
-apply_interval(Time, M, F, A) ->
-    req(apply_interval, {Time, self(), {M, F, A}}).
-
--spec send_interval(Time, Pid, Message) ->
-                           {'ok', TRef} | {'error', Reason} when
-      Time :: time(),
-      Pid :: pid() | (RegName :: atom()),
-      Message :: term(),
-      TRef :: tref(),
-      Reason :: term().
-send_interval(Time, Pid, Message) ->
-    req(apply_interval, {Time, Pid, {?MODULE, send, [Pid, Message]}}).
-
--spec send_interval(Time, Message) -> {'ok', TRef} | {'error', Reason} when
-      Time :: time(),
-      Message :: term(),
-      TRef :: tref(),
-      Reason :: term().
+          {'ok', TRef} | {'error', Reason}
+              when Time :: time(),
+                   Module :: module(),
+                   Function :: atom(),
+                   Arguments :: [term()],
+                   TRef :: tref(),
+                   Reason :: term().
+apply_interval(Time, M, F, A)
+  when ?valid_time(Time),
+       ?valid_mfa(M, F, A) ->
+    req(apply_interval, {system_time(), Time, self(), {M, F, A}});
+apply_interval(_Time, _M, _F, _A) ->
+    {error, badarg}.
+
+-spec send_interval(Time, Destination, Message) -> {'ok', TRef} | {'error', Reason}
+              when Time :: time(),
+                   Destination :: pid() | (RegName :: atom()) | {RegName :: atom(), Node :: node()},
+                   Message :: term(),
+                   TRef :: tref(),
+                   Reason :: term().
+send_interval(Time, Pid, Message)
+  when ?valid_time(Time),
+       is_pid(Pid) ->
+    req(apply_interval, {system_time(), Time, Pid, {?MODULE, send, [Pid, Message]}});
+send_interval(Time, RegName, Message)
+  when ?valid_time(Time),
+       is_atom(RegName) ->
+    req(apply_interval, {system_time(), Time, RegName, {?MODULE, send, [RegName, Message]}});
+send_interval(Time, Dest = {RegName, Node}, Message)
+  when ?valid_time(Time),
+       is_atom(RegName),
+       is_atom(Node) ->
+    req(apply_interval, {system_time(), Time, Dest, {?MODULE, send, [Dest, Message]}});
+send_interval(_Time, _Pid, _Message) ->
+    {error, badarg}.
+
+-spec send_interval(Time, Message) -> {'ok', TRef} | {'error', Reason}
+              when Time :: time(),
+                   Message :: term(),
+                   TRef :: tref(),
+                   Reason :: term().
 send_interval(Time, Message) ->
     send_interval(Time, self(), Message).
 
--spec cancel(TRef) -> {'ok', 'cancel'} | {'error', Reason} when
-      TRef :: tref(),
-      Reason :: term().
-cancel(BRef) ->
-    req(cancel, BRef).
-
--spec sleep(Time) -> 'ok' when
-      Time :: timeout().
+-spec cancel(TRef) -> {'ok', 'cancel'} | {'error', Reason}
+              when TRef :: tref(),
+                   Reason :: term().
+cancel({once, TRef})
+  when is_reference(TRef) ->
+    _ = erlang:cancel_timer(TRef, [{info, false}]),
```

Comment:
How come you use don't use `{async, true}` here as well? There is no guarantee that the timer has not already been triggered when `timer:cancel()` has returned, so the user must be prepared for a possible timeout message anyway.

`[{async, true}, {info, false}]` gives a real performance gain in case the timer was created on another scheduler while the gain from `[{info, false}]` is minimal (avoids creating an integer).

Reply:
You do have a point :) I'll see to it after lunch.

Reply:
The combo `[{async, true}, {info, false}]` is also not effected by the `[{info, false}]` bug which is nice :-)

Reply:
I don't know if that was me... if so, I just forgot or was distracted or something.

Reply:
Added `{async, true}` in last commit.

Reply:
Oh, I just remembered... I did that, and for a reason. I also realized now that I didn't do it properly, though...

The scenario concerns a `send_after` to `self()` timer. When `cancel` is called on such a timer and the timer has not expired yet, it will be stopped and no message will be sent. If it has expired however, the message should also be in the inbox, so it can be instantly flushed out with `receive ... after 0 -> ... end`. For that reason a synchronous (implicit `{async, false}`) call to `erlang:cancel_timer`, which will return only when the timer has really been cancelled (or doesn't exist any more).

I noticed that this falls short when the timer has just expired and told the timer server to execute the send, but the timer server has not done so yet, so a message may still arrive later and be missed by the flush. So we can as well use `{async, true}` for that matter.

Question now is, what shall we do? ;) Keep it like it is now (with `{async, true}`), or do we have to guarantee what the previous implementation (`{async, false}`) tried to achieve?

Reply:
> Oh, I just remembered... I did that, and for a reason.

@rickard-green and I discussed this. We arrived at the following:

With the current timer implementation, a message from the timer process can arrive after canceling the timer, but it is guaranteed that the message can be discarded like this:

```
receive
   <pattern that matches the timer message> -> ok
after 0 ->
   ok
end
```

With the new timer implementation, that guarantee no longer holds.

We think it is necessary to retain the guarantee that a `receive ... after 0 -> ok` will discard the timer message. To do that, the timer needs to be canceled in the timer server process (synchronously using `erlang:cancel_timer/1`). That is, the client process send a `cancel` request to the timer server process, the timer server process cancels the timer, and sends a reply to to client process.

Reply:
I forgot:

The timer server also needs to discard its timer message after canceling the timer. (`timer:apply_after()` should use `erlang:start_timer()` instead of `erlang:send_after()` to include the TimerRef in the timer message.)

Reply:
Hm... synchronous cancelling and flushing out a possible apply message in the timer server is a bit unfortunate. When many frequent cancels happen, the performance of the timer server may degrade.

What do you think of this alternative approach:
* The `erlang:send_after` timer is still cancelled in the client process, but with `[{async, false}, {info, true}]`.
  * If the result is `<some-integer>`, the timer still existed and has been stopped before it expired. Nothing will be sent to the timer server, so we are done here.
  * If the result is `false`, we know that the timer has expired and that the apply message is in the timer server and cannot be stopped any more; but we _don't_ know if the apply message has already been processed. So the client sends a call like `sync` to the timer server, to which the timer server will simply reply with `ok`.
When the server receives such a `sync` message, all apply messages that came before it (from a just-expired timer) will have been processed because they arrived in the mailbox before the `sync`. When the reply arrives at the client, we know the action we could not stop any more has been performed, and so cancel is done here.

This approach would keep the timer server almost entirely out of the picture, all it may have to do is unconditionally reply to the `sync` message, and even that will only happen if the erlang timer could not be cancelled by the client.

Reply:
@Maria-12648430 Sounds like a great solution!

Reply:
@Maria-12648430 I have replaced `%` with `%%` in your updated cancel code, re-indented, and force-pushed the updated commit. I intend to merge this pull request on Monday if everything goes well in our daily builds.

Reply:
Oops, sorry, forgot about that 😅

Reply:
Sorry I responded to fast :-( The BIF timer message is not guaranteed to be delivered to the server before the sync message is delivered since they originate from two different entities (see [signal doc](https://erlang.org/doc/reference_manual/processes.html#signal-delivery)) which will be a problem since the timer action might trigger after the cancel operation has returned.

Reply:
That complicates things... But I have an idea how it can be done. @Maria-12648430, meet on IRC tomorrow to discuss?

Reply:
Not sure when I'm available... just mail me the details?

Reply:
Ok :)

Reply:
@rickard-green @bjorng I discussed with @juhlig and implemented it like this:
* one-shot timers are now started from the timer server much like interval timers, and like those, they are registered in the timer table
* only timeout messages from timers in that table cause an action. Timeouts of one-shot timers cause removal from the table, timeouts of interval timers cause rescheduling and update.
* when a timer, one-shot or interval, is cancelled, it is removed from the table (if it still exists) and the erlang timer is stopped (if it still exists), so either the action has been performed already, or it will never be performed, no matter if and when the timeout message from the timer arrives

What do you think?

(And I hope I did the formatting right this time =^^=)

Reply:
@Maria-12648430 Looks good!

I guess the special case where the timer is aimed at the caller of `timer:send_after()` could be optimized by handling setup/cancel in the client though.

Reply:
Hm, I don't see how, off the top of my head ^^; What do you suggest? =^^=

Reply:
Are this what you are suggesting?

```erlang
send_after(Time, Pid, Msg) when Pid =:= self() ->
    TRef = erlang:send_after(Time, Pid, Msg),
    {ok, {detached, TRef}};
...
cancel({detached, TRef}) ->
    _ = erlang:cancel_timer(TRef),
    {ok, cancel};
```

If yes, two things:
* there is a tiny change in behavior, don't know if it matters: if the timer server gets stopped or crashes, naturally, all timers will be stopped; _except_ those detached timers, they will keep running.
* why not do that for _all_ local pids (`when node(Pid) =:= node()`), not just for `self()`?

Reply:
I'm wrong, it doesn't matter if it is aimed at yourself or not, but rather if it is a `send_after()` operation or not. Since the `tref()` type is opaque, no one is allowed to peek into it and the implementation is allowed to change it. Using a different tag, like for example `once_message`, when setting up a `send_after()` timer instead of `once`, `cancel()` can distinguish between `send_after()` timers and other timers and cancel the appropriate way.

Reply:
I didn't see your latest reply before replying. About the behavior change, yes that is true. Without having put much thought at it: Perhaps the timer server should be placed under the kernel supervisor so that it brings down the whole node if crash? Whan do you think @bjorng?

Have to do some errands, so I wont be able to reply for a few hours.

Reply:
Ugh, might the timer server be stopped :-( thats annoying...

Reply:
@rickard-green It already is under `kernel_safe_sup`, or rather, it starts itself under it... and it can't be stopped in a easy straightforward way. So that's ok then?

Reply:
@rickard-green well, it is not meant to be stopped. So that's probably ok ;)

Reply:
@rickard-green I think it's reasonable to place the timer server under the Kernel supervisor. I can't see any reason for why starting and stopping of the timer server should be supported.

Reply:
I take that as a yes :)

Reply:
Should it be moved into the `kernel` application then?

Reply:
@juhlig We avoid moving modules from stdlib to kernel because it could cause various problems. The dets server is now under the kernel supervisor, but the dets modules are still in stdlib.

Reply:
Last commit adds optimized `send_after` (and according `cancel`) for local pids.

Reply:
@bjorng I see, less work :) Then just change the `kernel_safe_sup` in `timer.erl` to `kernel_sup`?
Actually, `kernel_sup` is where it is placed if the timer server is started via kernel env param `start_timer` set to `true`. But otherwise, when the timer server is started dynamically, it is placed under `kernel_safe_sup`.

Reply:
@Maria-12648430 yes, the module doesn't even have a `stop()` function, so I guess it should be rather uncontroversial change to not support stop of the server :-) @juhlig yes, using `kernel_sup` instead of `kernel_safe_sup` should be enough.

Reply:
Good morning :) With the last commit, the timer server is started under `kernel_sup`.

@rickard-green @bjorng if you think everything is fine now, tell me and I'll squash it all together and augment the commit message accordingly.

Reply:
Looks good to *me*. I suggest that you wait for Rickard's answer before you squash.

Reply:
Of course :)

Reply:
Looks good to me too!